### PR TITLE
TSCH support for the Simplelink platform

### PR DESCRIPTION
**Pull request by atiselsts**
This PR is based on the initial work of @hseokman (#1038) It adds support of TSCH for the IEEE mode. 

**Changes:**
The code is quite long as it separates RAT (radio timer) stuff in a new .C file and makes sure RTC and RAT are always synchronized. There is some refactoring of the `ieee_mode` and `prop_mode` data structures to allow to access them from a base class ( `simplelink_radio_mode_t`) pointer.

**Testing**
CC2652R1 nodes, CC2650 nodes, CC1312R1 nodes and a CC1310 node have been tested.

*People are welcome to try it out and report problems or lack of them!* 

Depends on #1295

### CODE REVIEW

**Comment by atiselsts** [comment]
Now verified that basic functionality also works on CC2650 and that transmission works on CC1310 (had only single CC1310 node that could be flashed via JTAG).

**Comment by hseokman** [comment]
I tried your branch, and it works on CC1312R. I guess your change worked to keep RAT and RTC in sync. However, there must be some default timeout set for 2.4Ghz that is faster than expected in the sub Ghz network. The device joins the network and falls off network after some time and it continues the same behavior. 
Do you think RAT timer still needs manual start and stop? I thought the TI radio driver automatically handles that.

**Comment by atiselsts** [comment]
@hseokman - thanks for trying out. Do you have logs that you could share?

I have a couple ideas what could go wrong:
1) The `receiving_packet` function in IEEE mode is now behaving better, but I did not touch the prop mode code. It would be one of the first things to look.
2) The maximum waiting time in `transmit` code  might not be sufficient to transmit the whole packet in 50 kbps mode.

I have a question about the change you did in `netstack_sched_fs()`. The previous code tried to start the FS multiple times. After your change, just a single attempt is made. Should we add back the loop?

I have another question about `EVENTS_CMD_DONE` - you removed this macro from a couple places and replaced with checking the `CMD_STATUS`. Are you sure the change does not break anything?

I'm also on gitter, feel free to write there for a quicker turnaround.

**Comment by hseokman** [comment]
I attached log from the device.
I have not had chance to look at why it is leaving the network. I think one of default time out period is short that the device is leaving the network.
[tsch_log.txt](https://github.com/contiki-ng/contiki-ng/files/4594722/tsch_log.txt)

I copied receiving_packet from cc26xx-cc13xx prop mode to simplelink prop mode.

I don't understand why netstack_sched_fs() required a loop. I don't know if syncth_error is expected that cmd_fs needs to be passed until the error is cleared.
The main reason I changed this and other areas is that RF_runCmd cannot run in the interrupt handler. TSCH calls radio layer from rtimer handler which makes them called in the interrupt handler. I guess this is for the scheduling at the precise time, so I changed them to RF_scheduleCmd and wait for the command is executed.

**Comment by szilardkantor** [comment]
Could you explain please why is this neccessary when there has already been this pull?https://github.com/contiki-ng/contiki-ng/pull/648

**Comment by atiselsts** [comment]
The PR you linked to is for the CC26x0/CC13x0 platform. TI SimpleLink [is a newer platform](https://github.com/contiki-ng/contiki-ng/wiki/Platform-simplelink ) that adds support for CC2652R1, CC1312R1 and other devices, as well as runs on the older ones.

**Comment by szilardkantor** [comment]
I couldn't make it associate with 2 cc1310 launchpads. 
The node logs:
[INFO: TSCH      ] scan: received packet (20 bytes) on channel 25
[WARN: TSCH      ] scan: dropping packet, timestamp too far from current time 66520498 49579540
Is it supposed to send 20 bytes?

**Comment by atiselsts** [comment]
@szilardkantor thanks for the report, the `bAppendTimestamp` was not included in the RF settings for CC1310 radios. I have updated the PR.

**Comment by atiselsts** [comment]
Testing on my side is done. I'll keep watch on this for a few weeks to see if anyone tries this out reports any remaining problem. If all is good, then I'll remove the "work-in-progress" label and consider is a candidate for merging.

**Comment by andresblanco06** [comment]
@atiselsts What is the status on this? any help needed?

**Comment by atiselsts** [comment]
@andresblanco06
All four main board types have been tested (cc13x0, cc13x2, cc26x0, cc26x2) and interoperability with the cc26x0-cc13x0 platform was tested as well.

Just fixed the remaining outstanding issues I was aware of - in IEEE mode, 124 and 125 byte payload packets were not received correctly, and IEEE mode with CSMA was not working well.

Doing larger scale tests would be good, but it isn't normally required before merging PRs in the Contiki-NG develop branch. Anyway, if you can help / are aware of suitable testbeds let me know.

File: os/net/mac/tsch/tsch.c [hunk 1]
```diff
@@ -928,7 +928,7 @@ PROCESS_THREAD(tsch_send_eb_process, ev, data)
         }
       }
     }
-    if(tsch_current_eb_period > 0) {
+    if(0 && tsch_current_eb_period > 0) {
```
Comment (alexrayne) [inline os/net/mac/tsch/tsch.c:931]: you turn off EB here. How do you compensate it? don't see in other code, how EB sends.

**Comment by atiselsts** [comment]
Removed the work-in-progress status on this one. Ready for review, depends on #1295 which should be handled before this can be merged.

**Comment by sunilassao** [comment]
Have you verified the 200Kbps mode?

**Comment by atiselsts** [comment]
@sunilassao the PR only adds support to 250 kbps mode in the 2.4 GHz range and 50 kbps mode in the 868/915 MHz range.
All other radio modes are untested and, and modes with different data rates (5 kbps, 2.5 kbps etc) are going to need different TSCH timings.

**Pull request merged**